### PR TITLE
Compatibility with bokeh layout refactor

### PR DESCRIPTION
**Pull request by philippjfr**
This PR achieves compatibility with the layout refactor which was merged into bokeh dev and will land in bokeh 1.1.0. It will give us control over the actual plot frame dimensions (excluding the axes, colorbars, legends) and provide control over the plot aspect.

- [x] Adds support for ``aspect`` and ``data_aspect`` in bokeh (https://github.com/pyviz/holoviews/issues/335)
- [x] Fixes support for sizing modes in bokeh
- [x] Adds lower level control over reactive layouts using min/max_width/height and width/height_policy options
- [x] Ensure data aspect is maintained when updating plot
- [x] Ensure datetime and categorical axes warn if data_aspect set
- [x] Test on bokeh server
- [x] Check no events bounce on datashade
- [x] Update documentation
- [x] Update tests
- [x] Add tests

Fixes https://github.com/pyviz/holoviews/issues/3026
Fixes https://github.com/pyviz/holoviews/issues/2992

### CODE REVIEW

**Comment by philippjfr** [comment]
The main thing to settle here is how we should handle the switch from widths/heights which include axes/colorbars/legends to frame widths which apply only to the plot itself. 
The original behavior makes it very easy to get multiple plots to align nicely so it still has uses and we don't want to mess with everyones existing layouts without at least some warning so I think there are two options I can think of:

1) Add an option to switch between frame size and overall size (and for 2.0 we could consider switching the default to the frame width/height)
2) Add separate frame_width/frame_height options which override the standard width/height

**Comment by philippjfr** [comment]
I'm going to make an attempt at explaining the ``width``, ``height``, ``aspect`` and ``data_aspect`` behavior and explain how they interact with the plot **dimensions** and **ranges** particularly as it relates to responsive behavior and updating plots.

## width/height/frame_height/frame_width

width/height: Specifies the dimensions of the plot as a whole including legends/axes/colorbars/titles
frame_width/frame_height: Specifies the dimensions of just the plot frame

These options take absolute precedence over **everything** if a fixed height or width is specified by a user no other option including ``aspect``, ``data_aspect`` or ``responsive`` will have an effect (and will warn). A ``frame_width``/``frame_height`` will take precedence over a regular width/height.

### Aspect

The ``aspect`` specifies the ratio between the width and height dimensions of the plot. If specified this options takes absolute precedence over the **dimensions** of the plot but has no effect on the plot **ranges**. The only exception is ``aspect='equal'`` which is equivalent to ``data_aspect=1``. Aspect will only take effect if either or both and height and width are undefined.

### Data aspect

The ``data_aspect`` specifies the scaling between the x- and y-axis ranges. If specified this option will scale both the plot **ranges** and **dimensions** unless an ``aspect`` value overrides the plot dimensions.

### Responsive

In ``responsive`` mode the plot size rescales to the size of its container depending on the other options that were specified. ``responsive`` maps to bokeh's sizing mode as follows:

* ``stretch_both``: Stretches plot to maximum available width and height if no width, height or aspect are provided.
* ``stretch_width``: Stretches plot to maximum available width if no width and no aspect are provided
* ``stretch_height``: Stretches plot to maximum available height if no height and no aspect are provided
* ``scale_both``: Scales both width and height of plot while maximizing the width of the plot, if aspect is defined but no fixed width/height are. 
* ``scale_width``: Scales both width and height of plot while maximizing the width of the plot, if aspect is defined but no fixed width/height are and if ``responsive='width'`` (rarely used).
* ``scale_height``: Scales both width and height of plot while maximizing the width of the plot, if aspect is defined but no fixed width/height are and if ``responsive='width'`` (rarely used). 

There is one major caveat when combining a data aspect and responsive mode, in this case we cannot fully guarantee that a specified data aspect is maintained. This is because in responsive mode we don't have control over the plot frame dimensions just the overall plot aspect. We may be able to alleviate this by writing some JS code that maintains equal aspect in the plot ranges or by dynamically syncing the plot frame extents.

## Updating Plots

The other major issue here is maintaining a fixed aspect while a plot is updating. We do not want a plot resizing when you update it (in fact I could not even get that working), so the only way to achieve a fixed ``data_aspect`` is adjusting the ranges when new data comes in that has a different aspect. This works well but suffers from the same issue described above when responsive is enabled.

**Comment by jbednar** [comment]
I know we discussed all of the above comments at some point, but my brain starts to hurt when trying to really think deeply about them now. That said, as far as I can see that all sounds like what we had discussed and seems like the best approach in each case.

**Comment by jlstevens** [comment]
I agree with Jim that this sounds just like we discussed. From the description, it sounds like everything behaves the way I would hope, except maybe for the caveat about `data_aspect`.

**Comment by ea42gh** [comment]
Feature request: plot_size option for GridSpace should support tuples

File: holoviews/plotting/bokeh/util.py [hunk 1]
```diff
@@ -148,27 +156,166 @@ def compute_plot_size(plot):
...
+        return width, height
     elif isinstance(plot, (Plot, DataTable, Spacer)):
-        width, height = plot.width, plot.height
-    return width, height
+        return plot.width, plot.height
+    else:
+        return 0, 0
+
+
+def compute_layout_properties(
```
Comment (jlstevens) [inline holoviews/plotting/bokeh/util.py]: That is one long function indeed! Glad it is now split out from the rest.

File: holoviews/plotting/bokeh/plot.py [hunk 2]
```diff
@@ -40,33 +38,8 @@ class BokehPlot(DimensionedPlot):
...
-        width available, while maintaining the original aspect ratio.
-
-        "scale_height" elements will responsively resize to fit to the
-        height available, while maintaining the original aspect ratio.
-
-        "scale_both" elements will responsively resize to for both the
-        width and height available, while maintaining the original
-        aspect ratio.""")
+    css_classes = param.List(default=[], doc="""
+        CSS classes to apply to the layout.""")
```
Comment (jlstevens) [inline holoviews/plotting/bokeh/plot.py]: As far as I can tell there are no documented examples of how to use this plot option... (but maybe there is a diff I haven't expanded in this PR somewhere)

File: holoviews/plotting/bokeh/plot.py [hunk 3]
```diff
@@ -463,6 +445,51 @@ class CompositePlot(BokehPlot):
     to such a plot.
     """
 
+    sizing_mode = param.ObjectSelector(default=None, objects=[
+        'fixed', 'stretch_width', 'stretch_height', 'stretch_both',
+        'scale_width', 'scale_height', 'scale_both', None], doc="""
```
Comment (jlstevens) [inline holoviews/plotting/bokeh/plot.py]: I guess it is worth exposing the bokeh semantics even if users should rarely need to set this explicitly (in which case I would expect it to override everything else)

Comment (jlstevens) [reply holoviews/plotting/bokeh/plot.py]: As this is CompositePlot I see this is only for containers...

File: holoviews/plotting/bokeh/util.py (hunk 1 above)
Comment (jlstevens) [reply holoviews/plotting/bokeh/util.py]: I suspect some simplification could be done here but that doesn't have to hold up this PR...

File: holoviews/plotting/bokeh/util.py [hunk 4]
```diff
@@ -148,27 +156,166 @@ def compute_plot_size(plot):
...
+                       '\'square\'.')
+
+    return ({'aspect_ratio': aspect_ratio,
+             'aspect_scale': aspect_scale,
+             'match_aspect': match_aspect,
+             'sizing_mode' : sizing_mode},
+            {'frame_width' : frame_width,
+             'frame_height': frame_height,
+             'plot_height' : height,
+             'plot_width'  : width})
```
Comment (jlstevens) [inline holoviews/plotting/bokeh/util.py]: It isn't clear to me from the docstring why two dictionaries of settings are being returned here...

**Comment by jlstevens** [comment]
@philippjfr Looks great! Only had one minor request (docstring improvement) and other than that, I just want a way to remember to remove `-c bokeh/label/dev` once the new bokeh version is released.

Happy to merge once the small change is pushed (no need to wait on the tests this time).

File: holoviews/plotting/bokeh/plot.py (hunk 2 above)
Comment (philippjfr) [reply holoviews/plotting/bokeh/plot.py]: I'll add a small example to the bokeh guide, it's mostly useful for server apps I would think but adding CSS in a notebook is straightforward enough.

File: holoviews/plotting/bokeh/util.py (hunk 4 above)
Comment (philippjfr) [reply holoviews/plotting/bokeh/util.py]: I'll add a docstring.

File: holoviews/plotting/bokeh/plot.py (hunk 2 above)
Comment (philippjfr) [reply holoviews/plotting/bokeh/plot.py]: Tbh, I'm probably going to remove it, I haven't been able to come up with a useful example. It makes sense for Row and Column containers once we add them but at the plot level messing with css just doesn't seem useful.

**Comment by jlstevens** [comment]
Great thanks! Merging.

**Pull request merged**